### ISSUE TITLE
CNI went to panic after setting ANNOTATE_POD_IP to true

### ISSUE THREAD

**Issue by Downager** [aws/amazon-vpc-cni-k8s#1973]
**What happened**:

When I set ANNOTATE_POD_IP to true, the CNI pod will go to CrashloopBackOff, and the pod log showing observed a panic.
 
```
{"level":"info","ts":"2022-04-18T08:38:29.273Z","caller":"entrypoint.sh","msg":"Validating env variables ..."}
{"level":"info","ts":"2022-04-18T08:38:29.274Z","caller":"entrypoint.sh","msg":"Install CNI binaries.."}
{"level":"info","ts":"2022-04-18T08:38:29.294Z","caller":"entrypoint.sh","msg":"Starting IPAM daemon in the background ... "}
{"level":"info","ts":"2022-04-18T08:38:29.295Z","caller":"entrypoint.sh","msg":"Checking for IPAM connectivity ... "}
{"level":"info","ts":"2022-04-18T08:38:31.303Z","caller":"entrypoint.sh","msg":"Retrying waiting for IPAM-D"}
{"level":"info","ts":"2022-04-18T08:38:33.311Z","caller":"entrypoint.sh","msg":"Retrying waiting for IPAM-D"}
{"level":"info","ts":"2022-04-18T08:38:33.329Z","caller":"entrypoint.sh","msg":"Copying config file ... "}
{"level":"info","ts":"2022-04-18T08:38:33.333Z","caller":"entrypoint.sh","msg":"Successfully copied CNI plugin binary and config file."}
{"level":"info","ts":"2022-04-18T08:38:33.334Z","caller":"entrypoint.sh","msg":"Foregrounding IPAM daemon ..."}
E0418 08:38:42.505198      11 runtime.go:78] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
goroutine 562 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x55c7a3daaf60, 0x55c7a4068130})
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xa})
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:48 +0x75
panic({0x55c7a3daaf60, 0x55c7a4068130})
    /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*IPAMContext).AnnotatePod.func1()
    /go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/ipamd.go:1922 +0x16f
k8s.io/client-go/util/retry.OnError.func1()
    /go/pkg/mod/k8s.io/client-go@v0.20.2/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x12)
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:211 +0x72
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x0)
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:399 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x0, 0xc0008b01f0)
    /go/pkg/mod/k8s.io/client-go@v0.20.2/util/retry/util.go:50 +0xa8
k8s.io/client-go/util/retry.RetryOnConflict(...)
    /go/pkg/mod/k8s.io/client-go@v0.20.2/util/retry/util.go:104
github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*IPAMContext).AnnotatePod(0x0, {0xc0008f4020, 0x12}, {0xc0008180a8, 0x40}, {0x55c7a2ba8684, 0x7}, {0x0, 0xc000a96278})
    /go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/ipamd.go:1915 +0x145
github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*server).DelNetwork(0xc000b358a8, {0xc0000da790, 0x55c7a1872726}, 0xc0000da790)
    /go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/rpc_handler.go:276 +0xc32
github.com/aws/amazon-vpc-cni-k8s/rpc._CNIBackend_DelNetwork_Handler({0x55c7a3debe60, 0xc000b358a8}, {0x55c7a40d73d8, 0xc000a96240}, 0xc00024a420, 0x0)
    /go/src/github.com/aws/amazon-vpc-cni-k8s/rpc/rpc.pb.go:731 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002971e0, {0x55c7a4107ca0, 0xc00014c000}, 0xc000022200, 0xc00067f0e0, 0x55c7a50b5ed8, 0x0)
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:1082 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc0002971e0, {0x55c7a4107ca0, 0xc00014c000}, 0xc000022200, 0x0)
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:1405 +0x9f3
google.golang.org/grpc.(*Server).serveStreams.func1.1()
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:746 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:744 +0xef
panic: assignment to entry in nil map [recovered]
    panic: assignment to entry in nil map

goroutine 562 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xa})
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x55c7a3daaf60, 0x55c7a4068130})
    /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*IPAMContext).AnnotatePod.func1()
    /go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/ipamd.go:1922 +0x16f
k8s.io/client-go/util/retry.OnError.func1()
    /go/pkg/mod/k8s.io/client-go@v0.20.2/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x12)
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:211 +0x72
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x0)
    /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:399 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x0, 0xc0008b01f0)
    /go/pkg/mod/k8s.io/client-go@v0.20.2/util/retry/util.go:50 +0xa8
k8s.io/client-go/util/retry.RetryOnConflict(...)
    /go/pkg/mod/k8s.io/client-go@v0.20.2/util/retry/util.go:104
github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*IPAMContext).AnnotatePod(0x0, {0xc0008f4020, 0x12}, {0xc0008180a8, 0x40}, {0x55c7a2ba8684, 0x7}, {0x0, 0xc000a96278})
    /go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/ipamd.go:1915 +0x145
github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*server).DelNetwork(0xc000b358a8, {0xc0000da790, 0x55c7a1872726}, 0xc0000da790)
    /go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/rpc_handler.go:276 +0xc32
github.com/aws/amazon-vpc-cni-k8s/rpc._CNIBackend_DelNetwork_Handler({0x55c7a3debe60, 0xc000b358a8}, {0x55c7a40d73d8, 0xc000a96240}, 0xc00024a420, 0x0)
    /go/src/github.com/aws/amazon-vpc-cni-k8s/rpc/rpc.pb.go:731 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002971e0, {0x55c7a4107ca0, 0xc00014c000}, 0xc000022200, 0xc00067f0e0, 0x55c7a50b5ed8, 0x0)
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:1082 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc0002971e0, {0x55c7a4107ca0, 0xc00014c000}, 0xc000022200, 0x0)
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:1405 +0x9f3
google.golang.org/grpc.(*Server).serveStreams.func1.1()
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:746 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
    /go/pkg/mod/google.golang.org/grpc@v1.29.0/server.go:744 +0xef
Stream closed EOF for kube-system/aws-node-g5qvf (aws-node)
```

**What you expected to happen**:
All pods are in the running status.

**How to reproduce it (as minimally and precisely as possible)**:
Set `ANNOTATE_POD_IP` to true, and create a pod with empty annotations.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.19.15 (Self-hosted)
- CNI Version: v1.11.0
- OS (e.g: `cat /etc/os-release`): Flatcar Container Linux by Kinvolk 2905.2.6 (Oklo)
- Kernel (e.g. `uname -a`): 5.10.75-flatcar

**Comment by jayanthvn** [comment]
Hi. Have you enabled pod patch permission for aws-node service account?

**Comment by Downager** [comment]
I found the quick reproduce way:
Set ANNOTATE_POD_IP to true, and create a pod with empty annotations.

**Comment by jayanthvn** [comment]
Yeah was about ask, since the crash is seen in delete network was the pod created before setting the annotation?

**Comment by jayanthvn** [comment]
I will check on how it can be handled.

**Comment by jayanthvn** [comment]
Thanks for a quick fix. PR is merged to master and will pick it up for 1.11.1 release.
